File: demo_code.py
#!/usr/bin/env python3
"""
Demo Python file for testing LSP integration.
Contains functions and classes to test go-to-definition, find-references, and hover.
"""

from typing import List, Optional
from dataclasses import dataclass


@dataclass
class User:
    """User entity with authentication capabilities."""
    id: int
    name: str
    email: str
    is_active: bool = True


class AuthenticationError(Exception):
    """Raised when authentication fails."""
    pass


class UserService:
    """Service for managing user operations and authentication."""
    
    def __init__(self):
        self.users: List[User] = []
        self.current_user: Optional[User] = None
    
    def authenticate(self, email: str, password: str) -> User:
        """
        Authenticate a user with email and password.
        
        Args:
            email: User's email address
            password: User's password
            
        Returns:
            Authenticated User object
            
        Raises:
            AuthenticationError: If credentials are invalid
        """
        # Mock authentication logic
        user = self.find_user_by_email(email)
        if not user:
            raise AuthenticationError(f"User not found: {email}")
        
        # In real implementation, verify password hash
        if password == "wrong":
            raise AuthenticationError("Invalid password")
        
        self.current_user = user
        return user
    
    def find_user_by_email(self, email: str) -> Optional[User]:
        """Find user by email address."""
        return next((user for user in self.users if user.email == email), None)
    
    def create_user(self, name: str, email: str) -> User:
        """Create a new user."""
        user = User(
            id=len(self.users) + 1,
            name=name,
            email=email
        )
        self.users.append(user)
        return user


def main():
    """Main function demonstrating user service usage."""
    service = UserService()
    
    # Create users
    user1 = service.create_user("Alice", "alice@example.com")
    user2 = service.create_user("Bob", "bob@example.com")
    
    # Authenticate user
    try:
        authenticated_user = service.authenticate("alice@example.com", "password123")
        print(f"Authenticated: {authenticated_user.name}")
    except AuthenticationError as e:
        print(f"Authentication failed: {e}")


if __name__ == "__main__":
    main()

File: demo_code.ts
// TypeScript demo file for LSP testing
import { User, UserService, AuthenticationError } from './user-service';

interface LoginRequest {
    email: string;
    password: string;
}

class AuthenticationManager {
    private userService: UserService;
    private currentUser: User | null = null;

    constructor(userService: UserService) {
        this.userService = userService;
    }

    async authenticate(request: LoginRequest): Promise<User> {
        try {
            // Find user by email first
            const user = await this.userService.findUserByEmail(request.email);
            if (!user) {
                throw new AuthenticationError(`User not found: ${request.email}`);
            }

            // Validate credentials (mock logic)
            if (request.password === 'wrong') {
                throw new AuthenticationError('Invalid password');
            }

            this.currentUser = user;
            return user;
        } catch (error) {
            if (error instanceof AuthenticationError) {
                throw error;
            }
            throw new AuthenticationError('Authentication failed');
        }
    }

    getCurrentUser(): User | null {
        return this.currentUser;
    }

    logout(): void {
        this.currentUser = null;
    }
}

// Usage example
async function main() {
    const userService = new UserService();
    const authManager = new AuthenticationManager(userService);

    try {
        const user = await authManager.authenticate({
            email: 'alice@example.com',
            password: 'password'
        });
        
        console.log(`Authenticated: ${user.name}`);
        console.log(`Current user: ${authManager.getCurrentUser()?.name}`);
        
        authManager.logout();
        console.log(`After logout: ${authManager.getCurrentUser()}`);
    } catch (error) {
        if (error instanceof AuthenticationError) {
            console.error(`Authentication failed: ${error.message}`);
        } else {
            console.error(`Unexpected error: ${error}`);
        }
    }
}

main();